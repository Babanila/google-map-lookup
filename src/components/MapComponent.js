import React from 'react'
import { LoadScript, GoogleMap, Marker, InfoWindow } from '@react-google-maps/api'
import { cx, css } from 'emotion'
import { ParagraphDisplay } from './ReusableComponents'
import { initialMarker, getLocationDetails } from '../utils/helpers'

const containerStyle = {
  width: '100%',
  maxWidth: '500px',
  height: '500px'
}

const infoDiv = css`
  height: auto;
  background: #ffffff;
  color: #000000;
  border: 1px solid #ccc;
  padding: 5px;
  display: flex;
  flex-direction: column;
  line-height: 0.05em;
`

const infoSpan = css`
  color: #36a7fe;
  font-style: italic;
`

function MapComponent({ addressData }) {
  const [currentPosition, setCurrentPosition] = React.useState({})
  const [markerDetails, setMarkerDetails] = React.useState(initialMarker)
  const [showDetails, setShowDetails] = React.useState(false)
  const { results, status } = addressData

  const setInputLocation = () => {
    const { lat, lng } = results[0].geometry.location
    const currentPosition = { lat: parseInt(lat), lng: parseInt(lng) }
    setCurrentPosition(currentPosition)
  }

  React.useEffect(() => {
    navigator.geolocation.getCurrentPosition(setInputLocation)
  }, [])

  const handleClick = () => {
    const mark = getLocationDetails(results[0])
    setMarkerDetails({ ...mark })
    setShowDetails(true)
  }
  if (status !== 'OK') alert('Missing Location Data, please click the "Back" button')

  const defaultCenter = {
    lat: 52.520008,
    lng: 13.404954
  }

  return (
    <LoadScript googleMapsApiKey={process.env.API_KEY}>
      <GoogleMap
        mapContainerStyle={containerStyle}
        zoom={10}
        center={currentPosition.lat ? currentPosition : defaultCenter}
      >
        {currentPosition.lat && <Marker position={currentPosition} onClick={handleClick} />}

        {showDetails ? (
          <InfoWindow position={currentPosition} onCloseClick={() => setShowDetails(false)}>
            <div className={cx(infoDiv)}>
              <ParagraphDisplay
                leftOutput="Id"
                rightOutput={markerDetails.placeId}
                rightOutputStyles={infoSpan}
              />
              <ParagraphDisplay
                leftOutput="PostCode"
                rightOutput={markerDetails.postCode}
                rightOutputStyles={infoSpan}
              />
              <ParagraphDisplay
                leftOutput="Area"
                rightOutput={markerDetails.area}
                rightOutputStyles={infoSpan}
              />
              <ParagraphDisplay
                leftOutput="State"
                rightOutput={markerDetails.state}
                rightOutputStyles={infoSpan}
              />
              <ParagraphDisplay
                leftOutput="Country"
                rightOutput={markerDetails.country}
                rightOutputStyles={infoSpan}
              />
              <ParagraphDisplay
                leftOutput="Latitude"
                rightOutput={markerDetails.lat}
                rightOutputStyles={infoSpan}
              />
              <ParagraphDisplay
                leftOutput="Longitude"
                rightOutput={markerDetails.lng}
                rightOutputStyles={infoSpan}
              />
            </div>
          </InfoWindow>
        ) : null}
      </GoogleMap>
    </LoadScript>
  )
}

export default React.memo(MapComponent)
