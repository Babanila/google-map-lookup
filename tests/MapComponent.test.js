import React from 'react'
import renderer from 'react-test-renderer'
import { mount } from 'enzyme'
import { MemoryRouter } from 'react-router-dom'
import createGoogleMapsMock from 'jest-google-maps-mock'
import MapComponent from '../src/components/MapComponent'

describe('Check MapComponent Component', () => {
  let sampleData, googleMaps, process, props

  beforeEach(() => {
    googleMaps = createGoogleMapsMock()
    process = { env: { API_KEY: 'dummy-api-key-for-testing' } }
    props = { process }
    sampleData = {
      results: [
        {
          formatted_address: '90453 NÃ¼rnberg, Deutschland',
          geometry: {
            bounds: {
              northeast: { lat: 49.4012891, lng: 11.06465 },
              southwest: { lat: 49.35221, lng: 10.9925109 }
            },
            location: {
              lat: 49.3757248,
              lng: 11.0346054
            },
            location_type: 'APPROXIMATE',
            viewport: {
              northeast: { lat: 49.4012891, lng: 11.06465 },
              southwest: { lat: 49.35221, lng: 10.9925109 }
            }
          },
          place_id: 'ChIJ26TczgNRn0cRwEbF9DXaHhw',
          types: ['postal_code']
        }
      ],
      status: 'OK'
    }
  })

  afterEach(() => {
    jest.clearAllMocks()
  })

  it('should renders component without crashing', () => {
    const tree = renderer.create(<MapComponent addressData={sampleData} />).toJSON()
    expect(tree).toMatchSnapshot()
  })

  it('should render MapComponent component without data', () => {
    const MapComponentWrapper = mount(
      <MemoryRouter initialentries="{['/']}">
        <MapComponent addressData={sampleData} {...props} />
      </MemoryRouter>
    )

    const actualDiv = MapComponentWrapper.find('MemoryRouter')
    const actualDiv1 = actualDiv.find('Router')
    const actualComponent = actualDiv1.find('LoadScript')
    const childrenComponent = actualComponent.children()

    expect(MapComponent).toBeDefined()
    expect(actualDiv.length).toEqual(1)
    expect(actualDiv1.length).toEqual(1)
    expect(actualComponent.length).toEqual(1)
    expect(childrenComponent.length).toEqual(2)
    expect(childrenComponent.find('div').length).toEqual(2)
    expect(childrenComponent.find('div').at(1).childAt(0).text()).toEqual('Loading...')
  })

  it('should render MapComponent component with data', () => {
    const setCurrentPosition = React.useState

    jest.spyOn(React, 'useState').mockImplementationOnce(() =>
      setCurrentPosition({
        lat: 49.3757248,
        lng: 11.0346054
      })
    )

    const MapComponentWrap = mount(
      <MemoryRouter initialentries="{['/']}">
        <MapComponent addressData={{ ...sampleData }} />
      </MemoryRouter>
    )

    new googleMaps.Map(MapComponentWrap)
    const actualLoadScript = MapComponentWrap.find('LoadScript')

    expect(actualLoadScript.length).toEqual(1)
    expect(googleMaps.Map).toHaveBeenCalledTimes(1)
    expect(googleMaps.Map.mock.instances.length).toBe(1)
    expect(googleMaps.Map).toHaveBeenLastCalledWith(MapComponentWrap)
  })
})
