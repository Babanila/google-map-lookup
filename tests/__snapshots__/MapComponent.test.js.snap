// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Check MapComponent Component should renders component without crashing 1`] = `
Array [
  <div />,
  <div>
    Loading...
  </div>,
]
`;
